datalad/distribution/publish.py

```diff
@@ -75,33 +126,59 @@ def _publish_dataset(ds, remote, refspec, paths, annex_copy_options):
-        lgr.debug("Testing for changes with respect to '%s'", refspec)
-        # we have a matching branch on the other side
-        diff = ds.repo.repo.commit().diff(
-            refspec.replace(
-                'refs/heads/',
-                'refs/remotes/{}/'.format(remote)),
-            paths=paths)
     else:
-        # there was no tracking branch, check the push target
-        active_branch = ds.repo.get_active_branch()
-        refspec = active_branch
-        if active_branch in ds.repo.repo.remotes[remote].refs:
-            # we know some remote state -> diff
-            lgr.debug("Testing matching branch to '%s' at '%s' for differences",
-                      active_branch, remote)
-            diff = ds.repo.repo.commit().diff(
-                ds.repo.repo.remotes[remote].refs[active_branch],
-                paths=paths)
+        if refspec:
+            remote_branch_name = refspec[11:] \
+                if refspec.startswith('refs/heads/') \
+                else refspec
         else:
-            lgr.debug("No matching branch to '%s' at '%s'",
-                      active_branch, remote)
+            # there was no tracking branch, check the push target
+            remote_branch_name = ds.repo.get_active_branch()
+
+        if remote_branch_name in ds.repo.repo.remotes[remote].refs:
+            lgr.debug("Testing for changes with respect to '%s' of remote '%s'",
+                      remote_branch_name, remote)
+            current_commit = ds.repo.repo.commit()
+            remote_ref = ds.repo.repo.remotes[remote].refs[remote_branch_name]
+            if paths:
+                # if there were custom paths, we will look at the diff
+                lgr.debug("Since paths provided, looking at diff")
+                diff = current_commit.diff(
+                    remote_ref,
+                    paths=paths
+                )
+            else:
+                # if commits differ at all
+                lgr.debug("Since no paths provided, comparing commits")
+                diff = current_commit != remote_ref.commit
+        else:
+            lgr.debug("Remote '%s' has no branch matching %r. Will publish",
+                      remote, remote_branch_name)
             # we don't have any remote state, need to push for sure
             diff = True
 
+    # # remote might be set to be ignored by annex, or we might not even know yet its uuid
+    # annex_ignore = ds.config.getbool('remote.{}.annex-ignore'.format(remote), None)
+    # annex_uuid = ds.config.get('remote.{}.annex-uuid'.format(remote), None)
+    # if not annex_ignore:
+    #     if annex_uuid is None:
+    #         # most probably not yet 'known' and might require some annex
+    knew_remote_uuid = None
+    if isinstance(ds.repo, AnnexRepo):
+        try:
+            ds.repo.get_wanted(remote)  # could be just checking config.remote.uuid
+            knew_remote_uuid = True
+        except CommandError:
+            knew_remote_uuid = False
```

Comment:
this is the point I want to get improved, but not sure if would get time any time soon.... so probably will merge as is for now